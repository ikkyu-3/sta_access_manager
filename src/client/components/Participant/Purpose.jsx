// @flow
import type { ParticipantPurposeProps } from "@types";

import * as React from "react";
import { MEET_UP, STUDY, WORK, CIRCLE } from "@constants";
import styles from "./style.css";

function Purpose(props: ParticipantPurposeProps): React.Node {
  const { purpose } = props;
  const { columnPurpose, border, work, study, meetUp, circle } = styles;

  let background: string = "";
  let value: string = "";

  if (purpose === MEET_UP) {
    background = meetUp;
    value = "勉強会";
  }
  if (purpose === STUDY) {
    background = study;
    value = "自習";
  }
  if (purpose === WORK) {
    background = work;
    value = "仕事";
  }
  if (purpose === CIRCLE) {
    background = circle;
    value = "サークル";
  }

  return (
    <div className={`${columnPurpose} ${border} ${background}`}>{value}</div>
  );
}
export default Purpose;
