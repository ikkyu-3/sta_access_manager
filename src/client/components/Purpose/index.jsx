// @flow
import type { PurposeProps } from "@types";

import * as React from "react";
import styles from "./style.css";

class Purpose extends React.Component<PurposeProps> {
  componentDidMount() {
    this.props.checkId();
  }

  componentWillUnmount() {
    this.props.removeId();
  }

  render() {
    const { onPurposeButtonClick } = this.props;
    const { container, meetup, work, study, circle, message } = styles;

    return (
      <div className={container}>
        <button id="meetup" className={meetup} onClick={onPurposeButtonClick}>
          勉強会
        </button>
        <button id="work" className={work} onClick={onPurposeButtonClick}>
          仕事
        </button>
        <button id="study" className={study} onClick={onPurposeButtonClick}>
          自習
        </button>
        <button id="circle" className={circle} onClick={onPurposeButtonClick}>
          サークル
        </button>
        <div className={message}>目的を選択して下さい</div>
      </div>
    );
  }
}
export default Purpose;
