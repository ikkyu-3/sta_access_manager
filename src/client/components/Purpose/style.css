@charset "utf-8";
.container {
  display: grid;
  position: relative;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
}

.container > button {
  width: 100%;
  color: gray;
  margin: 0;
  padding: 0;
  font-size: 28px;
  font-weight: bold;
  border: none;
  box-sizing: border-box;
  outline: none;
}

.meetup {
  border-right: 2px solid lightgray !important;
  border-bottom: 2px solid lightgray !important;
  border-radius: 5px 0 0 0;
  background-color: #f8f189;
}

.work {
  border-left: 2px solid lightgray !important;
  border-bottom: 2px solid lightgray !important;
  border-radius: 0 5px 0 0;
  background-color: #bbfeff;
}

.study {
  border-right: 2px solid lightgray !important;
  border-top: 2px solid lightgray !important;
  border-radius: 0 0 0 5px;
  background-color: #bcf3cb;
}

.circle {
  border-left: 2px solid lightgray !important;
  border-top: 2px solid lightgray !important;
  border-radius: 0 0 5px 0;
  background-color: #f4d8f5;
}

.message {
  display: flex;
  position: absolute;
  width: 250px;
  height: 60px;
  left: calc(50% - 125px);
  top: calc(50% - 30px);
  justify-content: center;
  align-items: center;
  color: gray;
  font-size: 20px;
  font-weight: bold;
  background-color: white;
  border: 4px solid lightgray;
  border-radius: 10px;
  box-sizing: border-box;
}
