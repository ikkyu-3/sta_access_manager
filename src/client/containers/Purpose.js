// @flow
import type { State, Dispatch } from "@types";

import { connect } from "react-redux";
import { push } from "react-router-redux";

import { requestStart, requestEnd, removeId } from "@actions";
import Purpose from "@components/Purpose";
import {
  API_POST_ENTRY,
  TOP_PATH,
  COMPLETION_ENTRANCE_PATH,
  ERROR_BAD_REQUEST_PATH,
  ERROR_INTERNAL_SERVER_ERROR_PAHT,
  MEET_UP,
  STUDY,
  WORK,
  CIRCLE,
  UNKNOWN
} from "@constants";
import createApiUrl from "@modules/createApiUrl";
import request from "@modules/request";

const mapStateToProps = (state: State) => ({ state });

const mapDispatchToProps = (dispatch: Dispatch) => ({
  dispatch,
  removeId: () => {
    dispatch(removeId());
  }
});

const getPurpose = (purpose: string): string => {
  switch (purpose) {
    case "meetup":
      return MEET_UP;
    case "study":
      return STUDY;
    case "work":
      return WORK;
    case "circle":
      return CIRCLE;
    default:
      return "";
  }
};

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...dispatchProps,
  ...ownProps,
  checkId: () => {
    if (!/^[0-9a-f]{16}$/.test(stateProps.state.id)) {
      dispatchProps.dispatch(removeId());
      dispatchProps.dispatch(push(TOP_PATH));
    }
  },
  onPurposeButtonClick: async (
    e: SyntheticEvent<HTMLButtonElement>
  ): Promise<void> => {
    if (!(e.target instanceof window.HTMLButtonElement)) return;

    const purpose = getPurpose(e.target.id);
    if (purpose === UNKNOWN) return;

    const { id } = stateProps.state;

    const { dispatch } = dispatchProps;
    const { method, path } = API_POST_ENTRY;
    const url = createApiUrl(path);
    const data = { id, purpose };

    // 通信処理
    try {
      dispatch(requestStart());
      await request({ url, method, data });
      dispatch(requestEnd());

      dispatch(push(COMPLETION_ENTRANCE_PATH));
    } catch (error) {
      dispatch(requestEnd());

      if (error.status === 500) {
        dispatch(push(ERROR_INTERNAL_SERVER_ERROR_PAHT));
      } else {
        dispatch(push(ERROR_BAD_REQUEST_PATH));
      }
    }
  }
});

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(
  Purpose
);
