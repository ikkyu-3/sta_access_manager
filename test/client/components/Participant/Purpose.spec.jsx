import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Purpose from "@components/Participant/Purpose";

describe("Participant.Purpose.jsxのテスト", () => {
  describe("スナップショット", () => {
    it("正しいレンダリング", () => {
      const tree = renderer.create(<Purpose purpose="STUDY" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    it("purposeにSTUDYを指定した時、「自習」と表示される", () => {
      const purpose = shallow(<Purpose purpose="STUDY" />);
      expect(purpose.hasClass("study")).toBeTruthy();
      expect(purpose.text("自習")).toBeTruthy();
    });

    it("purposeにMEET_UPを指定した時、「勉強会」と表示される", () => {
      const purpose = shallow(<Purpose purpose="MEET_UP" />);
      expect(purpose.hasClass("meetUp")).toBeTruthy();
      expect(purpose.text("勉強会")).toBeTruthy();
    });

    it("purposeにWORKを指定した時、「仕事」と表示される", () => {
      const purpose = shallow(<Purpose purpose="WORK" />);
      expect(purpose.hasClass("work")).toBeTruthy();
      expect(purpose.text("仕事")).toBeTruthy();
    });

    it("purposeにcircleを指定した時、「サークル」と表示される", () => {
      const purpose = shallow(<Purpose purpose="CIRCLE" />);
      expect(purpose.hasClass("circle")).toBeTruthy();
      expect(purpose.text("サークル")).toBeTruthy();
    });

    it("purposeに上記以外の値を指定した時、表示されない", () => {
      const purpose = shallow(<Purpose purpose="test" />);
      expect(purpose.prop("className")).toBe("columnPurpose border ");
      expect(purpose.text()).toBe("");
    });
  });
});
