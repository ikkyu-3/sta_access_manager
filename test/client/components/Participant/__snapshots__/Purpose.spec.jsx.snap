// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Participant.Purpose.jsxのテスト スナップショット 正しいレンダリング 1`] = `
<div
  className="columnPurpose border study"
>
  自習
</div>
`;
